src/nwb_conversion_tools/tools/roiextractors/roiextractors.py

```diff
@@ -138,27 +142,44 @@ def add_devices(nwbfile: NWBFile, metadata: dict) -> NWBFile:
     return nwbfile
 
 
-def _add_imaging_plane(nwbfile: NWBFile, metadata=dict) -> NWBFile:
+def _create_imaging_plane_from_metadata(nwbfile: NWBFile, imaging_plane_metadata: dict) -> ImagingPlane:
 
-    metadata_copy = deepcopy(metadata)
-    add_devices(nwbfile=nwbfile, metadata=metadata_copy)
+    device_name = imaging_plane_metadata["device"]
+    imaging_plane_metadata["device"] = nwbfile.devices[device_name]
 
-    # Add the image plane
-    image_plane_metadata = metadata_copy["Ophys"]["ImagingPlane"][0]
+    imaging_plane_metadata["optical_channel"] = [
+        OpticalChannel(**metadata) for metadata in imaging_plane_metadata["optical_channel"]
+    ]
 
-    device_name = image_plane_metadata["device"]
-    image_plane_metadata["device"] = nwbfile.devices[device_name]
+    imaging_plane = ImagingPlane(**imaging_plane_metadata)
 
-    image_plane_metadata["optical_channel"] = [
-        OpticalChannel(**metadata) for metadata in image_plane_metadata["optical_channel"]
-    ]
+    return imaging_plane
+
+
+def add_imaging_plane(nwbfile: NWBFile, metadata=dict, imaging_plane_index: int = 0) -> NWBFile:
+
+    # Set the defaults and required infraestructure
+    metadata_copy = deepcopy(metadata)
+    default_metadata = get_default_ophys_metadata()
+    metadata_copy = dict_deep_update(default_metadata, metadata_copy, append_list=False)
+    add_devices(nwbfile=nwbfile, metadata=metadata_copy)
 
-    nwbfile.create_imaging_plane(**image_plane_metadata)
+    imaging_plane_metadata = metadata_copy["Ophys"]["ImagingPlane"][imaging_plane_index]
+    imaging_plane_name = imaging_plane_metadata["name"]
+    existing_imaging_planes = nwbfile.imaging_planes
+
+    if imaging_plane_name not in existing_imaging_planes:
+        imaging_plane = _create_imaging_plane_from_metadata(
+            nwbfile=nwbfile, imaging_plane_metadata=imaging_plane_metadata
+        )
+        nwbfile.add_imaging_plane(imaging_plane)
 
     return nwbfile
 
```

Comment:
The diff is doing a terrible job at displaying this but from line 141 to here the code displays the substitution of the old private method `_add_imaging_plane` for a new `add_imaging_plane` which adds the imaging plane to an nwb file and a second private function `_create_imaging_plane_from_metadata` which creates an image plane from the metadata when its name does not exist in the nwbfile yet.

Reply:
I like this approach :)

Reply:
Yes, I think that makes sense for this module. Having an auxiliary private function that creates the pynwb objects from the metadata and then a public tested function that adds this to the nwb file checking for edge cases and controlling parameters/options/etc.